src/firefly/js/core/ComponentCntlr.js

```diff
@@ -0,0 +1,177 @@
+    flux.process({type: HIDE_ALL_DIALOGS, payload: {}});
+}
+
+export function dispatchComponentStateChange(componentId, componentState) {
+    flux.process({type: COMPONENT_STATE_CHANGE, payload: {componentId, componentState}});
+}
+
+
+//============================ Utilities ============================
+//============================ Utilities ============================
+//============================ Utilities ============================
+
+export function isDialogVisible(dialogKey) {
+    return get(flux.getState()[DIALOG_OR_COMPONENT_KEY], [DIALOG_KEY,dialogKey,'visible'],false);
+}
+
+export function getDialogOwner(dialogKey) {
+    const d= get(flux.getState()[DIALOG_OR_COMPONENT_KEY], [DIALOG_KEY,dialogKey], {});
+    return (d.visible && d.ownerId) ? d.ownerId : null;
+}
+
+export function getComponentState(componentKey) {
+    return get(flux.getState()[DIALOG_OR_COMPONENT_KEY], [COMPONENT_KEY,componentKey], {});
+}
+
+/*
+   This controller handles small (one level deep) state of various application components,
+   such as dialogs, tab and collapsible panels, which are not a part of field group
+
+   Structure of the store:
+   dialogOrComponent:
+     dialog:
+        [dialogId]: {
+            visible
+            ownerId
+        }
+     component:
+        [componentId]: savedComponentState
+ */
+
+//======================================== Reducer =============================
+//======================================== Reducer =============================
+//======================================== Reducer =============================
+
+
+function reducer(state={[DIALOG_KEY]: {}, [COMPONENT_KEY]: {}}, action={}) {
+    if (!action.payload || !action.type) return state;
+    switch (action.type) {
+
+        case SHOW_DIALOG  :
+            return showDialogChange(state,action);
+
+        case HIDE_DIALOG  :
+            return hideDialogChange(state,action);
+
+        case HIDE_ALL_DIALOGS  :
+            return hideAllDialogsChange(state,action);
+
+        case COMPONENT_STATE_CHANGE  :
+            return changeComponentState(state,action);
+
+        default:
+            return state;
+    }
+}
+
+
+//============ private functions =================================
+//============ private functions =================================
+//============ private functions =================================
+
+const showDialogChange= function(state,action) {
+    var {dialogId,ownerId}= action.payload;
+    const dialogsState = state[DIALOG_KEY];
+
+    if (!dialogsState[dialogId] || !dialogsState[dialogId].visible || ownerId!==dialogsState[dialogId].ownerId) {
+        state = update(state,
+                        {[DIALOG_KEY] :
+                            {$merge: {
+                                [dialogId]: {
+                                    visible: true,
+                                    ownerId
+                                }
+                            }}});
+    }
```

Comment:
An Alternative:
`const clone = (obj,params={}) => Object.assign({},obj,params);`
Once defined:
```
     state = clone(state,
                       {[DIALOG_KEY] :
                           clone(state[DIALOG_KEY],
                               { [dialogId]: {
                                   visible: true,
                                   ownerId
                               } })});
```

It is more clear what a function is and what an object is.